### PR TITLE
Python ncnn with vulkan

### PR DESCRIPTION
**Pull request by Hideousmon**
Enable the usage of Vulkan with Python wheels. Details and troubleshooting can be found at [ncnn-gpu-pre-verification](https://github.com/Hideousmon/ncnn-gpu-pre-verification).

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=5020) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=5020) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=5020) it.</sub>

**Comment by Hideousmon** [comment]
I find that the Vulkan-Loader can be built and used well on some other platforms. There will be some improvements.

**Comment by Hideousmon** [comment]
The updated status:
|                       | x86_64 | i686 | x86  | AMD64 | ARM64 | universal2 | aarch64 | ppc64le | s390 |
| :-------------------: | :----: | :--: | :--: | :---: | :---: | :--------: | :-------: | :-------: | :----: |
| ubuntu (cp-manylinux) |   √    |  √   |  -   |   -   |   -   |     -      | √ | √ | √ |
| ubuntu (cp-musllinux) |   √    |  √   |  -   |   -   |   -   |     -      | √ | √ | √ |
|      ubuntu (pp)      |   √    |  √   |  -   |   -   |   -   |     -      | √ | - | - |
|     windows (pp)      |   -    |  -   |  -   |   √   |   -   |     -      | - | - | - |
|     windows (cp)      |   -    |  -   | cpu  |   √   |  cpu  |     -      | - | - | - |
|      macos (pp)       |   √    |  -   |  -   |   -   |   -   |     -      | - | - | - |
|      macos (cp)       |   √    |  -   |  -   |   -   |  cpu  |    cpu     | - | - | - |

**Comment by Hideousmon** [comment]
Find that the building process goes well when resetting -G "Visual Studio 16 2019" -A Win32 for Cmake on x86 Windows. Some improvements again. :sweat_smile:

**Comment by Hideousmon** [comment]
Finally, it works well on all the platforms. :relaxed:
|                       | x86_64 | i686 | x86  | AMD64 | ARM64 | universal2 | aarch64 | ppc64le | s390x |
| :-------------------: | :----: | :--: | :--: | :---: | :---: | :--------: | :-------: | :-------: | :----: |
| ubuntu (cp-manylinux) |   √    |  √   |  -   |   -   |   -   |     -      | √ | √ | √ |
| ubuntu (cp-musllinux) |   √    |  √   |  -   |   -   |   -   |     -      | √ | √ | √ |
|      ubuntu (pp)      |   √    |  √   |  -   |   -   |   -   |     -      | √ | - | - |
|     windows (pp)      |   -    |  -   |  -   |   √   |   -   |     -      | - | - | - |
|     windows (cp)      |   -    |  -   | √ |   √   |  √  |     -      | - | - | - |
|      macos (pp)       |   √    |  -   |  -   |   -   |   -   |     -      | - | - | - |
|      macos (cp)       |   √    |  -   |  -   |   -   |  √  |    √    | - | - | - |

Updated troubleshooting: https://github.com/Hideousmon/ncnn-gpu-pre-verification/blob/main/troubleshooting.md

**Comment by Hideousmon** [comment]
Updated release test for [a966dda](https://github.com/Tencent/ncnn/pull/5020/commits/a966ddab00d5ea2dad8781e03a017173674247a3):
https://github.com/Hideousmon/ncnn/actions/runs/6211745770

**Comment by Hideousmon** [comment]
Updated release test for [6d64456](https://github.com/Tencent/ncnn/pull/5020/commits/6d6445670a4e69a3b89dd874b44348cbfadc6aa3):
https://github.com/Hideousmon/ncnn/actions/runs/6348913870

**Comment by Hideousmon** [comment]
Updated release test for [7e9ed71](https://github.com/Tencent/ncnn/pull/5020/commits/7e9ed71efcaf481bd249e9b088a689ad9e71aeb9) [cibuildwheel@v2.15.0 -> v2.16.2]:
https://github.com/Hideousmon/ncnn/actions/runs/6445490723

File: .github/workflows/release-python.yml [hunk 1]
```diff
@@ -62,20 +70,222 @@ jobs:
...
+        CIBW_ENVIRONMENT: CMAKE_BUILD_PARALLEL_LEVEL=2 NCNN_VULKAN=ON VULKAN_SDK=$GITHUB_WORKSPACE/vulkansdk-macos-1.3.236.0/macOS
+          CMAKE_TOOLCHAIN_FILE=$GITHUB_WORKSPACE/toolchains/ios.toolchain.cmake PLATFORM=MAC_ARM64 ARCHS="arm64"
+          DEPLOYMENT_TARGET="11.0" ENABLE_BITCODE=OFF ENABLE_ARC=OFF ENABLE_VISIBILITY=OFF
+          OpenMP_C_FLAGS="-Xclang -fopenmp" OpenMP_CXX_FLAGS="-Xclang -fopenmp"
+          OpenMP_C_LIB_NAMES="libomp" OpenMP_CXX_LIB_NAMES="libomp"
+          OpenMP_libomp_LIBRARY="$DEVELOPER_DIR/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib/libomp.a"
+          Vulkan_INCLUDE_DIR=$GITHUB_WORKSPACE/vulkansdk-macos-1.3.236.0/MoltenVK/include
+          Vulkan_LIBRARY=$GITHUB_WORKSPACE/vulkansdk-macos-1.3.236.0/MoltenVK/dylib/macOS/libMoltenVK.dylib
+      with:
+        output-dir: wheelhouse
```
Comment (nihui) [inline .github/workflows/release-python.yml:288]: x86_64 package libvulkan.dylib while arm64/universal package libMotenVK.dylib
Can macos x86_64 package really work without moltenvk?

Comment (Hideousmon) [reply .github/workflows/release-python.yml]: Recheck the CI logging, and the repairing process didn't copy libMoltenVK.dylib for macos x86_64 indeed. The Vulkan_INCLUDE_DIR and Vulkan_LIBRARY are specified now in [b7d7e42](https://github.com/Tencent/ncnn/pull/5020/commits/b7d7e42c010dbda1c172532bebd9a34indeedd8cec5a05). The corresponding release test for [b7d7e42](https://github.com/Tencent/ncnn/pull/5020/commits/b7d7e42c010dbda1c172532bebd9a34indeedd8cec5a05) :
https://github.com/Hideousmon/ncnn/actions/runs/6584694265

It's worth noting that the previously built ncnn wheel can function properly with cpu, even without libMoltenVK.dylib on x86_64 macOS. The situation isn't too bad; it's quite promising, haha. 
(https://github.com/Hideousmon/ncnn-gpu-pre-verification/actions/runs/6212630410/job/16862967067 )

File: .github/workflows/release-python.yml [hunk 2]
```diff
@@ -62,20 +70,224 @@ jobs:
...
+    - name: Build wheels for macos arm64 and macos universal2
+      if: (matrix.os == 'macos-latest' && matrix.arch == 'arm64') ||
+        (matrix.os == 'macos-latest' && matrix.arch == 'universal2')
+      uses: pypa/cibuildwheel@v2.16.2
+      env:
+        CIBW_ARCHS_MACOS: ${{ matrix.arch }}
+        CIBW_BUILD: ${{ matrix.build }}
+        CIBW_BUILD_VERBOSITY: 1
+        CIBW_ENVIRONMENT: CMAKE_BUILD_PARALLEL_LEVEL=2 NCNN_VULKAN=ON VULKAN_SDK=$GITHUB_WORKSPACE/vulkansdk-macos-1.3.236.0/macOS
+          CMAKE_TOOLCHAIN_FILE=$GITHUB_WORKSPACE/toolchains/ios.toolchain.cmake PLATFORM=MAC_ARM64 ARCHS="arm64"
```
Comment (nihui) [inline .github/workflows/release-python.yml:282]: but universal build only contains arm64 arch  :(

Comment (Hideousmon) [reply .github/workflows/release-python.yml:282]: Some revision and tests are needed. I will have a try this weekend.

File: .github/workflows/release-python.yml (hunk 1 above)
Comment (Hideousmon) [reply .github/workflows/release-python.yml]: The ncnn wheel with libMoltenVK.dylib for x86_64 macOS can function properly:
```
(ncnn) zz@zzs-MacBook-Pro ncnnx86_64test % python yolov4-tiny-testbench.py
[0 Intel(R) Iris(TM) Plus Graphics]  queueC=0[1]  queueG=0[1]  queueT=0[1]
[0 Intel(R) Iris(TM) Plus Graphics]  bugsbn1=0  bugbilz=0  bugcopc=0  bugihfa=0
[0 Intel(R) Iris(TM) Plus Graphics]  fp16-p/s/a=1/1/1  int8-p/s/a=1/1/1
[0 Intel(R) Iris(TM) Plus Graphics]  subgroup=32  basic/vote/ballot/shuffle=1/1/1/1
[0 Intel(R) Iris(TM) Plus Graphics]  fp16-matrix-16_8_8/16_8_16/16_16_16=0/0/0
NCNN Init time: 515.338176 ms
NCNN Inference time: 109.97455200000006 ms
NCNN Detection time: 111.33643299999996 ms
```

File: .github/workflows/release-python.yml (hunk 2 above)
Comment (Hideousmon) [reply .github/workflows/release-python.yml:282]: "ios.toolchain.cmake" offers the MAC_UNIVERSAL options, it may be solved soon. 😄

Comment (Hideousmon) [reply .github/workflows/release-python.yml:281]: [7c1674b](https://github.com/Tencent/ncnn/pull/5020/commits/7c1674b5bf87826e129d557693807ef04d009388) is built with -target arm64-x86_64-apple-macosx11.0, it should be a 'real' universal2 wheel. The release test for it is:
https://github.com/Hideousmon/ncnn/actions/runs/6602816382
But I can not test it on a machine with gpu, when I try to install the universal2 wheel on my x86_64 macos (macOS 14.0, with pip==23.3.1), it is told that `ERROR: ncnn-1.0.20231022-cp38-cp38-macosx_11_0_universal2.whl is not a supported wheel on this platform`.

Comment (Hideousmon) [reply .github/workflows/release-python.yml:281]: Test it with a universal2 python on my x86_64 macos with gpu:
```
zz@zzs-MacBook-Pro bin % pip3 install ~/Downloads/ncnn-1.0.20231022-cp38-cp38-macosx_11_0_universal2.whl
zz@zzs-MacBook-Pro ncnnx86_64test % python3.8 yolov4-tiny-testbench.py
[0 Intel(R) Iris(TM) Plus Graphics]  queueC=0[1]  queueG=0[1]  queueT=0[1]
[0 Intel(R) Iris(TM) Plus Graphics]  bugsbn1=0  bugbilz=0  bugcopc=0  bugihfa=0
[0 Intel(R) Iris(TM) Plus Graphics]  fp16-p/s/a=1/1/1  int8-p/s/a=1/1/1
[0 Intel(R) Iris(TM) Plus Graphics]  subgroup=32  basic/vote/ballot/shuffle=1/1/1/1
[0 Intel(R) Iris(TM) Plus Graphics]  fp16-matrix-16_8_8/16_8_16/16_16_16=0/0/0
NCNN Init time: 521.049342 ms
NCNN Inference time: 112.29334700000004 ms
NCNN Detection time: 113.453374 ms
```
It is functional ! 😆

File: .github/workflows/release-python.yml [hunk 3]
```diff
@@ -289,6 +288,24 @@ jobs:
...
 
+    - name: Build wheels for macos universal2 
+      if: matrix.os == 'macos-latest' && matrix.arch == 'universal2'
+      uses: pypa/cibuildwheel@v2.16.2
+      env:
+        CIBW_ARCHS_MACOS: ${{ matrix.arch }}
+        CIBW_BUILD: ${{ matrix.build }}
+        CIBW_BUILD_VERBOSITY: 1
+        CIBW_ENVIRONMENT: CMAKE_BUILD_PARALLEL_LEVEL=2 NCNN_VULKAN=ON VULKAN_SDK=$GITHUB_WORKSPACE/vulkansdk-macos-1.3.236.0/macOS
+          CMAKE_TOOLCHAIN_FILE=$GITHUB_WORKSPACE/toolchains/ios.toolchain.cmake PLATFORM=MAC_UNIVERSAL ARCHS="arm64;x86_64"
```
Comment (nihui) [inline .github/workflows/release-python.yml:299]: ncnn不支持这种多个架构一起cmake，只能编译好最后 lipo 合起来
看编译log，会发现他只编译了 layer/arm/xxx 的东西，没有编译 layer/x86/xxx 的东西
简单的办法，暂时不提供 universal wheel ？

Comment (Hideousmon) [reply .github/workflows/release-python.yml]: 暂时先删掉了，等试试在setup.py的build_ext里lipo一下试试(再混个pr
删掉后的release测试：
https://github.com/Hideousmon/ncnn/actions/runs/6604184258

**Comment by nihui** [comment]
Thanks for your contribution !

**Pull request merged**